File: Data Structures & Algorithms/search-for-word/submission-0.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int ROWS = board.size(); 
        int COLS = board[0].size();

        for (int i = 0; i < ROWS; i++)
        {
            for (int j = 0; j < COLS; j++)
            {
                if (board[i][j] == word[0])
                {
                    unordered_set<string> visited;
                    if (dfs(i, j, ROWS, COLS, board, word, 0)) return true;
                }
            }
        }

        return false; 
    }

    bool dfs(int r, int c, int ROWS, int COLS, vector<vector<char>> &board, string word, int index)
    {
        if (index == word.length()) return true; 
        if (r < 0 || c < 0 || r >= ROWS || c >= COLS || (board[r][c] != word[index]) ) 
            return false; 
        
        char original_char = board[r][c];
        board[r][c] = '#';
    
        bool ans = dfs(r+1, c, ROWS, COLS, board, word, index+1) ||
        dfs(r, c+1, ROWS, COLS, board, word, index+1) ||
        dfs(r-1, c, ROWS, COLS, board, word, index+1) ||
        dfs(r, c-1, ROWS, COLS, board, word, index+1);

        board[r][c] = original_char;
        return ans; 
    }
};

File: Data Structures & Algorithms/valid-sudoku/submission-0.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        num = [{} for _ in range(9)]

        for i in range(len(board)):
            for j in range(len(board[i])):
                c = board[i][j]
                if c != '.':
                    if c in num[i] and num[i][c][0] > 0:
                        return False
                    if c in num[j] and num[j][c][1] > 0:
                        return False
                    if c in num[i//3*3 + j//3] and num[i//3*3 + j//3][c][2] > 0:
                        return False

                    if c not in num[i]:
                        num[i][c] = [0, 0, 0]
                    if c not in num[j]:
                        num[j][c] = [0, 0, 0]
                    if c not in num[i//3*3 + j//3]:
                        num[i//3*3 + j//3][c] = [0, 0, 0]

                    num[i][c][0] += 1
                    num[j][c][1] += 1
                    num[i//3*3 + j//3][c][2] += 1
        
        return True